src/compiler/utilities.ts

```diff
@@ -215,8 +215,37 @@ namespace ts {
 
     // Gets the nearest enclosing block scope container that has the provided node
     // as a descendant, that is not the provided node.
-    export function getEnclosingBlockScopeContainer(node: Node): Node {
-        let current = node.parent;
+    export function getEnclosingBlockScopeContainer(node: Node): Node;
+
+    /**
+      * Gets the nearest enclosing block scope container that has the provided node
+      * as a descendant, that is not the provided node.
+      * @param node The starting node
+      * @param getAncestorOrSelf A callback used to get the ancestor of the starting node, used 
+      * only when traversing the ancestors of the current node in the emitter.
+      * @param offset The offset in the node stack used to get the ancestor of the current node
+      * from the emitter's node stack.
+      */
+    export function getEnclosingBlockScopeContainer(node: Node, getAncestorOrSelf: (offset: number) => Node, offset: number): Node;
+
+    /**
+      * Gets the nearest enclosing block scope container that has the provided node
+      * as a descendant, that is not the provided node.
+      * @param node The starting node
+      * @param getAncestorOrSelf A callback used to get the ancestor of the starting node, used 
+      * only when traversing the ancestors of the current node in the emitter.
+      * @param offset The offset in the node stack used to get the ancestor of the current node
+      * from the emitter's node stack.
+      * @remarks
+      * The emitter tracks the parent of the current node as it descends into the source file, 
+      * so that we can properly emit synthesized nodes that may not have parent pointers.
+      * We can call `getAncestorOrSelf` with an offset that specifies how far back in the node's 
+      * ancestry to retrieve a parent, grandparent, etc. An offset of zero (0) refers to the 
+      * current node on the top of the node stack, an offset of one (1) refers to its parent, an 
+      * offset of two (2) refers to the grandparent, and so on.

```

Comment:
What if offset is `undefined` or `NaN`. If I am reading the implementation of `getEnclosingBlockScopeContainer`, we can call `getAncestorOfSelf(++offset)` with `offset` equal to `undefined`, so `getAncestorOfSelf` is called with `NaN`.